openmc/model/surface_composite.py

```diff
@@ -783,6 +784,41 @@ def _make_ccw(self, points):
 
         return points[::-1, :]
 
+    def _constrain_triangulation(self, points):
+        """Generate a constrained triangulation.
+
+        Parameters
+        ----------
+        points : np.ndarray (Nx2)
+            An Nx2 array of coordinate pairs describing the vertices. These
+            points represent a planar straight line graph.
+
+        Returns
+        -------
+        None
+        """
+
+        tri = Delaunay(points, qhull_options='QJ')
+        # Loop through the boundary edges of the polygon. If an edge is not
+        # included in the triangulation, break it into two line segments.
+        n = len(points)
+        new_pts = []
+        for i, j in zip(range(n), range(1, n +1)):
+            # If both vertices of any edge are not found in any simplex, insert
+            # a new point between them
+            if not any([i in s and j % n in s for s in tri.simplices]):
+                newpt = (points[i, :] + points[j % n, :]) / 2
+                new_pts.append((j, newpt))
+
+        # If all the edges are included in the triangulation set it, otherwise
+        # try again with additional points inserted on offending edges
+        if not new_pts:
+            self._tri = tri
+        else:
+            for i, pt in new_pts[::-1]:
+                points = np.insert(points, i, pt, axis=0)
+            self._constrain_triangulation(points)
```

Comment:
As mentioned I'd suggest constraining the recursion level here, and if over the max, throwing an error.